rgdps/usecases/friend_requests.py

```diff
@@ -0,0 +1,233 @@
+from rgdps.models.friend_request import FriendRequest
+from rgdps.models.user import User
+
+
+class FriendRequestResponse(NamedTuple):
+    request: FriendRequest
+    user: User
+
+
+class PaginatedFriendRequestResponse(NamedTuple):
+    requests: list[FriendRequestResponse]
+    total: int
+
+
+async def get_user(
+    ctx: Context,
+    user_id: int,
+    page: int = 0,
+    page_size: int = 10,
+    is_sender_user_id: bool = False,
+) -> PaginatedFriendRequestResponse | ServiceError:
+
+    requests = await repositories.friend_requests.from_user_id_paginated(
+        ctx,
+        user_id,
+        page,
+        page_size,
+        is_sender_user_id=is_sender_user_id,
+        include_deleted=False,
+    )
+
+    friend_request_responses = []
+    for request in requests:
+        # Swap the sender and recipient according to the from_sender_id flag
+        target_user_id = (
+            request.recipient_user_id if is_sender_user_id else request.sender_user_id
+        )
+        user = await repositories.user.from_id(ctx, target_user_id)
+
+        if user is None:
+            continue
+
+        friend_request_responses.append(
+            FriendRequestResponse(request, user),
+        )
+
+    friend_request_count = (
+        await repositories.friend_requests.get_user_friend_request_count(
+            ctx,
+            user_id,
+            is_sender_user_id=is_sender_user_id,
+        )
+    )
+
+    return PaginatedFriendRequestResponse(
+        friend_request_responses,
+        friend_request_count,
+    )
+
+
+async def mark_as_seen(
+    ctx: Context,
+    user_id: int,
+    request_id: int,
+) -> FriendRequest | ServiceError:
+    request = await repositories.friend_requests.from_id(ctx, request_id)
+
+    if request is None:
+        return ServiceError.FRIEND_REQUEST_NOT_FOUND
+
+    # This is only triggered by recipient.
+    if request.recipient_user_id != user_id:
+        return ServiceError.FRIEND_REQUEST_INVALID_OWNER
+
+    request = await repositories.friend_requests.update_partial(
+        ctx,
+        request_id,
+        seen_ts=datetime.now(),
+    )
+
+    if request is None:
+        return ServiceError.FRIEND_REQUEST_NOT_FOUND
+
+    return request
+
+
+async def accept(
+    ctx: Context,
+    sender_user_id: int,
+    recipient_user_id: int,
+    request_id: int,
+) -> ServiceError | None:
+    request = await repositories.friend_requests.from_id(ctx, request_id)
+
+    if request is None:
+        return ServiceError.FRIEND_REQUEST_NOT_FOUND
+
+    if (
+        request.sender_user_id != sender_user_id
+        or request.recipient_user_id != recipient_user_id
+    ):
+        return ServiceError.FRIEND_REQUEST_INVALID_OWNER
+
+    request = await repositories.friend_requests.update_partial(
+        ctx,
+        request.id,
+        deleted=True,
+    )
+
+    if request is None:
+        return ServiceError.FRIEND_REQUEST_NOT_FOUND
+    
+    if sender_user_id == recipient_user_id:
```

Comment:
this check can probably be done nearer to the start

Reply:
resolved in 0f37f98